src/v/cluster/rm_stm.cc

```diff
@@ -1663,6 +1663,7 @@ void rm_stm::apply_fence(model::producer_identity pid, model::record_batch b) {
 }
 
 ss::future<> rm_stm::do_apply(const model::record_batch& b) {
+    auto units = co_await _state_lock.hold_read_lock();
```

Comment:
i think this lock shouldn't be here. Apply should always be possible and the state should not be modified outside of apply method

Reply:
> the state should not be modified outside of apply method

I believe this is still true here (ie, the state is not modified outside of apply). This additional coordination is solely  to prevent undesirable interactions between prefix truncation (which resets all internal state) and concurrently executing apply fibers.